content/post/2019-11-21-readmes.Rmarkdown

```diff
@@ -0,0 +1,416 @@
+
+Now, a helpful way to still convey information efficiently when there are more than a few things is a good README _structure_. Besides, a good structure is important for READMEs of all sizes.
+
+### Glimpsing at README structure
+
+To assess README structures a bit, we first need to extract headers.
+
+
+```{r structure}
+get_headings <- function(xml) {
+  
+  if(is.null(xml)) {
+    return(NULL)
+  }
+  
+  xml %>%
+    xml2::xml_find_all("heading") -> headings
+  
+  list(tibble::tibble(text = xml2::xml_text(headings),
+                 position = seq_along(headings),
+                 level = xml2::xml_attr(headings, "level")))
+}
+
+structure <- readmes %>%
+  dplyr::group_by(repo, owner) %>%
+  dplyr::mutate(structure = get_headings(xml_readme[[1]])) %>%
+  dplyr::ungroup()
+```
+
+Here's the structure of the 42th README.
+
+```{r structureex}
+structure$structure[42][[1]]
+```
+
+We wrote an ugly long function to visualize the structure of any of the sampled READMEs, inspired by not ugly code in [`fs`](https://github.com/r-lib/fs/blob/master/R/tree.R).
+
+<details>
+<summary>Click to see the function.</summary>
+```{r ugly}
+
+pc <- function(...) {
+  paste0(..., collapse = "")
+}
+
+print_readme_structure <- function(structure){
+
+  structure$parent <- NA
+  for (i in seq_len(nrow(structure))) {
+    possible_parents <- structure$position[structure$level < structure$level[i]
+                                    & structure$position < structure$position[i]]
+    if (any(possible_parents)) {
+      structure$parent[i] <- max(possible_parents)
+    } else {
+      structure$parent[i] <- NA
+    }
+  }
+  
+  structure$parent[structure$level == min(structure$level)] <- 0
+  
+  
+  for (i in seq_len(nrow(structure))) {
+    if (structure$level[i] == 1) {
+      cat(structure$text[i])
+    } else {
+      
+      if(structure$position[i] == max(structure$position[structure$parent==structure$parent[i]])) {
+        firstchar <- cli:::box_chars()$l
+      } else {
+        firstchar <- cli:::box_chars()$j
+      }
+      
+      cat(
+        rep("  ", max(
+          0,
+          as.numeric(structure$level[i]) - 1
+        )),
+        pc(
+          firstchar,
+          pc(rep(
+            cli:::box_chars()$h,
+            max(
+              0,
+              as.numeric(structure$level[i]) - 1
+            )
+          ))
+        ), structure$text[i]
+      )
+    }
+    cat("\n")
+  }
+  
+}
+
+
+```
+</details>
+
+In practice,
+
+```{r r42}
+print_readme_structure(structure$structure[42][[1]])
+print_readme_structure(structure$structure[7][[1]])
+print_readme_structure(structure$structure[1][[1]])
```

Comment:
I think there is a bug in the tree drawing:

```
## A Teradata Backend for dplyr
##          ├─── Koji Makiyama (@hoxo-m)
##    ├─ 1 Overview
##    ├─ 2 Installation
##    ├─ 3 Details
##       ├── 3.1 Usage
##       ├── 3.2 Translatable functions
##          ├─── 3.2.1 lubridate friendly functions
##          ├─── 3.2.2 Treat Boolean
##          ├─── 3.2.3 to_timestamp()
##          └─── 3.2.4 cut()
##       └── 3.3 Other useful functions
##          └─── 3.3.1 blob_to_string()
##    └─ 4 Related work
```

Reply:
what would you expect here?

Reply:
The broken line on the left to continue down to "Related work".